File: lexic/front.cc
#include <cstdio>
#include <cctype>
#include <cstring>
#include <cstdlib>

enum CharClass {
  LETTER, DIGIT, DOT, UNKNOWN
};

enum Token {
  INT_LIT = 10,
  FLOAT_LIT,
  INDENT,
  ASSIGN_OP,
  ADD_OP,
  SUB_OP,
  MULT_OP,
  DIV_OP,
  LEFT_PAREN,
  RIGHT_PAREN,
  INVALID
};

int charClass;
char lexeme[100];
char nextChar;
int lexLen;
int nextToken;
FILE *in_fp;

void addChar();
void getChar();
void getNonBlank();
int lex();

int main() {
  if ((in_fp = fopen("front.in", "r")) == NULL) {
    printf("ERROR - cannot open front.in\n");
    exit(1);
  }
  getChar();
  do {
    lex();
  } while (nextToken != EOF);
}

int lookup(char ch) {
  switch (ch) {
    case '(':
      addChar();
      nextToken = LEFT_PAREN;
      break;
    case ')':
      addChar();
      nextToken = RIGHT_PAREN;
      break;
    case '+':
      addChar();
      nextToken = ADD_OP;
      break;
    case '-':
      addChar();
      nextToken = SUB_OP;
      break;
    case '*':
      addChar();
      nextToken = MULT_OP;
      break;
    case '/':
      addChar();
      nextToken = DIV_OP;
      break;
    case EOF:
      addChar();
      nextToken = EOF;
      break;
    default:
      addChar();
      nextToken = INVALID;
  }
  return nextToken;
}

void addChar() {
  if (lexLen <= 98) {
    lexeme[lexLen++] = nextChar;
    lexeme[lexLen] = 0;
  } else {
    printf("ERROR - lexeme is too long\n");
  }
}

void getChar() {
  if ((nextChar = getc(in_fp)) != EOF) {
    if (isalpha(nextChar))
      charClass = LETTER;
    else if (isdigit(nextChar))
      charClass = DIGIT;
    else if (nextChar == '.')
      charClass = DOT;
    else
      charClass = UNKNOWN;
  } else {
    charClass = EOF;
  }
}

void getNonBlank() {
  while (isspace(nextChar))
    getChar();
}

int lex() {
  lexLen = 0;
  getNonBlank();
  if (charClass == LETTER) {
    addChar();
    getChar();
    while (charClass == LETTER or charClass == DIGIT) {
      addChar();
      getChar();
    }
    nextToken = INDENT;
  } else if (charClass == DIGIT or charClass == DOT) {
    int dots = charClass == DOT, letters = 0;
    addChar();
    getChar();
    while (charClass == DIGIT or charClass == DOT or charClass == LETTER) {
      dots += charClass == DOT;
      letters += charClass == LETTER;
      addChar();
      getChar();
    }
    if (letters or dots > 1)
      nextToken = INVALID;
    else if (dots == 0)
      nextToken = INT_LIT;
    else
      nextToken = FLOAT_LIT;
  } else if (charClass == UNKNOWN) {
    lookup(nextChar);
    getChar();
  } else if (charClass == EOF) {
    nextToken = EOF;
    strcpy(lexeme, "EOF");
  }
  if (nextToken == INVALID) {
    printf("ERROR - invalid token %s\n", lexeme);
  } else {
    printf("Next token is: %d, Next lexeme is %s\n", nextToken, lexeme);
  }
  return nextToken;
}

File: lexic/Front.java
import java.io.BufferedReader;
import java.io.EOFException;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.lang.Character;
import java.lang.StringBuilder;

public class Front {
    private CharClass charClass;
    private StringBuilder lexeme;
    char nextChar;
    private Token nextToken;
    BufferedReader input;

    public static void main(String[] args) {
        FileReader fr;
        try {
            fr = new FileReader(new File("front.in"));
        } catch (java.io.FileNotFoundException e) {
            return;
        }
        Front f = new Front(new BufferedReader(fr));
        f.run();
    }

    public Front(BufferedReader input) {
        this.input = input;
        this.lexeme = new StringBuilder();
    }

    public void run() {
        getChar();
        do {
            lex();
        } while (nextToken != Token.EOF);
    }

    void lookup() {
        switch (nextChar) {
            case '(':
                addChar();
                nextToken = Token.LEFT_PAREN;
                break;
            case ')':
                addChar();
                nextToken = Token.RIGHT_PAREN;
                break;
            case '+':
                addChar();
                nextToken = Token.ADD_OP;
                break;
            case '-':
                addChar();
                nextToken = Token.SUB_OP;
                break;
            case '*':
                addChar();
                nextToken = Token.MULT_OP;
                break;
            case '/':
                addChar();
                nextToken = Token.DIV_OP;
                break;
            default:
                addChar();
                nextToken = Token.INVALID;
        }
    }

    void addChar() {
        lexeme.append(nextChar);
    }

    void getChar() {
        int nextInt;
        try {
            nextInt = input.read();
        } catch (IOException e) {
            charClass = CharClass.EOF;
            return;
        }
        nextChar = (char) nextInt;
        if (nextInt == -1) {
            charClass = CharClass.EOF;
            return;
        }
        if (Character.isAlphabetic(nextChar))
            charClass = CharClass.LETTER;
        else if (Character.isDigit(nextChar))
            charClass = CharClass.DIGIT;
        else if (nextChar == '.')
            charClass = CharClass.DOT;
        else
            charClass = CharClass.UNKNOWN;
    }

    void getNonBlank() {
        while (Character.isWhitespace(nextChar))
            getChar();
    }

    void lex() {
        lexeme = new StringBuilder();
        getNonBlank();
        if (charClass == CharClass.LETTER) {
            addChar();
            getChar();
            while (charClass == CharClass.LETTER || charClass == CharClass.DIGIT) {
                addChar();
                getChar();
            }
            nextToken = Token.IDENT;
        } else if (charClass == CharClass.DIGIT || charClass == CharClass.DOT) {
            int dots = (charClass == CharClass.DOT) ? 1 : 0, letters = 0;
            addChar();
            getChar();
            while (charClass == CharClass.DIGIT || charClass == CharClass.DOT ||
                    charClass == CharClass.LETTER) {
                dots += (charClass == CharClass.DOT) ? 1 : 0;
                letters += (charClass == CharClass.LETTER) ? 1 : 0;
                addChar();
                getChar();
            }
            if (letters > 0 || dots > 1)
                nextToken = Token.INVALID;
            else if (dots == 0)
                nextToken = Token.INT_LIT;
            else
                nextToken = Token.FLOAT_LIT;
        } else if (charClass == CharClass.UNKNOWN) {
            lookup();
            getChar();
        } else if (charClass == charClass.EOF) {
            nextToken = Token.EOF;
            lexeme.append("EOF");
        }

        if (nextToken == Token.INVALID)
            System.out.printf("ERROR - invalid token %s\n", lexeme.toString());
        else
            System.out.printf("Next token is: %s, Next lexeme is %s\n", nextToken, lexeme);
    }

}

enum CharClass {
    LETTER,
    DIGIT,
    DOT,
    UNKNOWN,
    EOF
};

enum Token {
    INT_LIT,
    FLOAT_LIT,
    IDENT,
    ASSIGN_OP,
    ADD_OP,
    SUB_OP,
    MULT_OP,
    DIV_OP,
    LEFT_PAREN,
    RIGHT_PAREN,
    INVALID,
    EOF
};
